### PR TITLE
FAI-694: Quarkus: runtime-tools-quarkus-extension: Make UI features pluggable

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.redhat.com/browse/FAI-694

This PR responds to two new `BuildItem`'s signalling when a back-end service is available.

They determine what UI components are visible in the DevUI console.

This is part of an ensemble:
- https://github.com/kiegroup/kogito-runtimes/pull/1818
- https://github.com/kiegroup/kogito-apps/pull/1176

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins retest this

**Comment by manstis** [comment]
Jenkins retest this

File: runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/audit.html [hunk 1]
```diff
@@ -31,13 +31,15 @@
...
-    container: document.getElementById("envelope-app"),
-    users: {info:userInfo.arrayRepresentation.raw},
-    dataIndexUrl: "{config:property('kogito.data-index.url')}/graphql",
-    trustyServiceUrl: "{config:property('kogito.trusty.http.url')}",
-    page: "Audit"
-  })
+    const devUI = RuntimeToolsDevUI.open({
+        container: document.getElementById("envelope-app"),
+        isProcessEnabled: {info:isProcessEnabled},
+        isTracingEnabled: {info:isTracingEnabled},
```
Comment (kelvah) [inline runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/audit.html:37]: You should make the same change inside [`ui-packages/packages/runtime-tools-dev-ui-webapp/resources/index.html`](https://github.com/kiegroup/kogito-apps/blob/main/ui-packages/packages/runtime-tools-dev-ui-webapp/resources/index.html#L39) to keep the `yarn dev` script working for the runtime ui-package.
I think the only option is to set them to `true` there.

File: runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html [hunk 2]
```diff
@@ -14,6 +14,7 @@
   ~ limitations under the License.
   -->
 
+{#if info:isProcessEnabled}
```
Comment (kelvah) [inline runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html:17]: I have a general question related to this card. Is it possible that both process and tracing are disabled?
In that case this card will be empty. Should we care to display a message eventually?

File: ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUIRoutes/DevUIRoutes.tsx [hunk 3]
```diff
@@ -14,63 +14,106 @@
...
-          counterfactualEnabled={false}
-          explanationEnabled={false}
-          containerConfiguration={{
-            pageWrapper: false,
-            serverRoot: trustyServiceUrl,
-            basePath: '/Audit',
-            excludeReactRouter: true,
-            useHrefLinks: false
-          }}
+      {isProcessEnabled && (
```
Comment (kelvah) [inline ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUIRoutes/DevUIRoutes.tsx]: Just nitpicking here.
Wrapping each route with the same condition is a bit repetitive. You could place all the process routes inside an array of JSX elements and render the array like `{isProcessEnabled && processRoutes}`.
wdyt?

**Review by kelvah — COMMENTED** [review]
Hi @manstis, I left only one major comment and other minor observations.

File: runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/audit.html (hunk 1 above)
Comment (manstis) [reply runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/audit.html:37]: Done. Thank-you. Good catch.

File: runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html (hunk 2 above)
Comment (manstis) [reply runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html:17]: The scenario you describe is definitely possible. 

I've added an extra section to `embedded.html` for when neither are enabled.

File: ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUIRoutes/DevUIRoutes.tsx (hunk 3 above)
Comment (manstis) [reply ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUIRoutes/DevUIRoutes.tsx]: Refactored.. I hope you like it?

File: runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html (hunk 2 above)
Comment (manstis) [reply runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html:17]: A classic example of my web-design expertise:

No extensions

Comment (pefernan) [reply runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html:17]: @manstis haha I'd prefer something like `No Kogito addons detected` but that works.

Comment (manstis) [reply runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html:17]: OK @pefernan @kelvah I've updated to this wording.

Comment (kelvah) [reply runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/resources/dev-templates/embedded.html:17]: :100:

File: ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUIRoutes/DevUIRoutes.tsx (hunk 3 above)
Comment (kelvah) [reply ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUIRoutes/DevUIRoutes.tsx]: Nice! Don't know why I was worrying about wrapping routes with react fragments. Perfectly fine!

**Review by kelvah — APPROVED** [review]

**Comment by manstis** [comment]
GHA keeps failing with:
```
2022-01-20T10:44:39.5237796Z [ERROR] /home/runner/work/kogito-apps/kogito-apps/kiegroup_kogito_apps/kogito-apps/runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/java/org/kie/kogito/runtime/tools/quarkus/extension/deployment/DevConsoleProcessor.java:[22,55] cannot find symbol
2022-01-20T10:44:39.5240831Z [ERROR]   symbol:   class TrustyServiceAvailableBuildItem
2022-01-20T10:44:39.5243477Z [ERROR]   location: package org.kie.kogito.quarkus.addons.common.deployment
2022-01-20T10:44:39.5251456Z [ERROR] /home/runner/work/kogito-apps/kogito-apps/kiegroup_kogito_apps/kogito-apps/runtime-tools-quarkus-extension-parent/runtime-tools-quarkus-extension-deployment/src/main/java/org/kie/kogito/runtime/tools/quarkus/extension/deployment/DevConsoleProcessor.java:[102,28] cannot find symbol
2022-01-20T10:44:39.5254481Z [ERROR]   symbol:   class TrustyServiceAvailableBuildItem
2022-01-20T10:44:39.5257956Z [ERROR]   location: class org.kie.kogito.runtime.tools.quarkus.extension.deployment.DevConsoleProcessor
2022-01-20T10:44:39.5260282Z [ERROR] -> [Help 2]
2022-01-20T10:44:39.5268146Z org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project runtime-tools-quarkus-extension-deployment: Compilation failure
```
It is (for whatever reason) failing to build the `kogito-runtimes` FAI-694 branch.

Jenkins is quite capable and happy...

**Review by pefernan — APPROVED** [review]
+1, thanks @manstis!

File: ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUILayout/tests/DevUILayout.test.tsx [hunk 4]
```diff
@@ -29,20 +29,124 @@ const ApolloClientMock = ApolloClient as jest.MockedClass<typeof ApolloClient>;
...
+    expect(
+      wrapper.exists('Link[data-ouia-navigation-name="forms-list-nav"]')
+    ).toBeFalsy();
+    expect(
+      wrapper.exists('Link[data-ouia-navigation-name="audit-nav"]')
+    ).toBeTruthy();
+  });
+
+  it('Snapshot test with Tracing disabled', () => {
+    // eslint-disable-next-line
```
Comment (sjamboro) [inline ui-packages/packages/runtime-tools-dev-ui-webapp/src/components/DevUI/DevUILayout/tests/DevUILayout.test.tsx]: Do we want to disable eslint?

**Review by sjamboro — CHANGES_REQUESTED** [review]
Code looks good. The code coverage is high. Jenkins build is green.

Unfortunately,  GHA (mvn v3.8.1) is failing with CNFE `org.kie.kogito.quarkus.addons.common.deployment.TrustyServiceAvailableBuildItem`. It seems like it is not a scope of this PR because that class was created by different PR ([#1818](https://github.com/kiegroup/kogito-runtimes/pull/1818) - kogito-runtime) which was approved but not merged  (@jiripetrlik @danielezonca @cristianonicolai can you take a look at it).

I am not able to build it locally. When I was using mvn v3.6.3 that class was missing in kogito-runtime jar
image - strange behaviour. When we merge it so we can break all GHA checks.

When I have updated to v3.8.4 (latest version) the jar contains that class. Unfortunately I was not able to build kogito-apps from different reasons (failed yarn run build:prod). I still investigate that problem.

@manstis I left comments about eslint and ts ignore (3 or 4 occurences in one test file). Can you remove it? I think that it is not a good idea ignore rules in the "final" code.

**Review by sjamboro — APPROVED** [review]

**Pull request merged**